index.js

```diff
@@ -7,256 +7,273 @@ var _ = require('lodash');
+     * @callback callback
+     */
+  EchoClient.prototype.queryAnalytics = function queryAnalytics(
+    token,
+    queryOperator,
+    queryParams,
+    useCache,
+    callback
+  ) {
     if (!token) {
-        throw new Error('Missing Persona token');
+      throw new Error('Missing Persona token');
     }
 
     if (!queryOperator) {
-        throw new Error('Missing Analytics queryOperator');
+      throw new Error('Missing Analytics queryOperator');
     }
 
     var validOperators = ['hits', 'average', 'sum', 'max', 'funnel'];
 
     if (validOperators.indexOf(queryOperator) === -1) {
-        throw new Error('Invalid Analytics queryOperator');
+      throw new Error('Invalid Analytics queryOperator');
     }
 
     if (!queryParams) {
-        throw new Error('Missing Analytics queryParams');
+      throw new Error('Missing Analytics queryParams');
     }
 
-    var constructQueryStringResponse = this._queryStringParams(queryParams);
+    var constructQueryStringResponse = queryStringParams(queryParams);
 
     if (!_.isEmpty(constructQueryStringResponse.errors)) {
-        this.error('One or more invalid analytics queryParams where supplied: ' + constructQueryStringResponse.errors.join());
-        throw new Error('Invalid Analytics queryParams');
+      error(
+        'One or more invalid analytics queryParams where supplied',
+        { err: constructQueryStringResponse.errors.join() }
+      );
+
+      throw new Error('Invalid Analytics queryParams');
     }
 
     var requestOptions = {
-        url: this.config.echo_endpoint + '/1/analytics/' + queryOperator + '?' + constructQueryStringResponse.queryString,
-        method: 'GET',
-        headers: {
-            'Content-Type': 'application/json',
-            'Authorization': 'Bearer ' + token
-        }
+      url: config.echo_endpoint + '/1/analytics/' + queryOperator + '?' + constructQueryStringResponse.queryString,
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: 'Bearer ' + token
+      }
     };
 
     if (useCache === false) {
-        requestOptions.headers['cache-control'] = 'none';
+      requestOptions.headers['cache-control'] = 'none';
     }
 
-    this.debug(JSON.stringify(requestOptions));
+    debug('request options', requestOptions);
 
-    var _this = this;
+    request.get(requestOptions, function onResp(err, response, rawBody) {
+      var status = response && response.statusCode
+        ? response.statusCode
+        : 0;
 
-    request.get(requestOptions, function(err, response, body) {
-        if (err) {
-            _this.error('[echoClient] queryAnalytics error: ' + JSON.stringify(err));
-            callback(err);
-        } else {
-            _this._parseJSON(body, callback);
-        }
+      if (err || parseInt(status / 100) !== 2) {
+        error('[echoClient] queryAnalytics error', { err: err, body: rawBody });
+        callback(err || 'invalid status code: ' + status);
```

Comment:
why `invalid status code`?  I'm not sure it makes sense to have that returned to the caller does it? What does that mean?

Also, the pattern used here is not used elsewhere in other `if (err || parseInt(status / 100) !== 2)` blocks, any reason?

Reply:
either there was a explicit error or the status code of the request was not as expected. What would you like the code to return if the status code was not of 2xx?

I don't understand which code you are refering to. Can you show me which line?

Reply:
What I am meaning is that in the code here, you are doing `callback(err || 'invalid status code: ' + status);` but in the code below for example, it's just `callback(err);`. I was asking why in this case you specifically return either the err OR a hardcoded message.

```
+    request.post(requestOptions, function onResp(err, response, body) {
 +      var statusCode = response && response.statusCode
 +        ? response.statusCode
 +        : 0;
 +
 +      if (err || parseInt(statusCode / 100) !== 2) {
 +        error('[echoClient] addEvents error', { err: err, body: body });
 +        callback(err);
 +      } else {
 +        callback(null, body);
 +      }
 +    });